### PR TITLE
Add UP and DOWN keys to scroll past commands

### LINKED ISSUE #77: Use up arrow and down arrow to scroll through past commands
Just like in command-line.

(inspired from `https://github.com/AY2324S2-CS2103T-F13-3/tp/issues/53`)

### PR DESCRIPTION
**Pull request by ChillinRage**
Closes #77 

Didn't use `ListIterator` since pressing up and down repeatedly will not change the command shown (due to how they implement `next` and` previous`), but it should toggle between 2 commands.

### CODE REVIEW

**Comment by ChillinRage** [comment]
I tried implementing test case for `commandBox`, but seems like I can't get `commandBox` itself to be displayed (unlike `helpWindow`).

Every method is set to private, and is called through the GUI.
`commandTextField.requestFocus()` then `robot.press(...)` wouldn't work either since the box is not a window.

**Review by E0735389 — CHANGES_REQUESTED** [review]
For testing this, a relatively simple way is to factor out the logic part into a `CommandHistoryModel`, which contains

* `addExecutedCommand(String command)`
* `String getPreviousCommandHistory()`
* `String getNextCommandHistory()`

and test only that. Of course the GUI is still untested but I suppose it's fine.

Design-wise, this also somewhat makes sense (later on you may want to write a web-based GUI or a terminal user interface using the same model? Why duplicate the code? There's nothing inherently JavaFX with the model)

Also, `bash` has this feature: assume you type a part of the current command `abc`, the previous command is `def`, you (accidentally) press `[↑]`, the command-line changes to `def`, then you press `[↓]`, the command should change **back** to `abc`.

Usability-wise, this makes sense (if the user accidentally press `[↑]`, there should be an easy way to undo it instead of making the user re-type the command again)



This also applies when the user is editing a **past** command and accidentally press `[↑]`.

Apart from that point, we need not imitate bash's behavior completely. (bash also has the feature that if you edit some past command, the edit *persists* across executions, this is not too important in my opinion)

**Comment by ChillinRage** [comment]
A bit of things I want to clarify.
Suppose that the current simplified history is `add`, `edit 1`, `delete 4`.
- User types `hel`, then press UP. > now showing `delete 4`
- User edits it to `delete 2` then press UP again. > now showing `edit 1`
- User edits it to `edit` then press DOWN. > now showing `delete 2`
- User press DOWN again, > now showing `hel`
- User edits it to `help` and enter.
- Now the history is back to `add`, `edit 1`, `delete 4`, `help`. 

Is that the gist of it? (In regards to what you're suggesting?)

I'll omit the "edit persists across executions" because in windows, pressing up and down will not save the current changes (i.e. current behavior of this `commandHistory`).

**Comment by E0735389** [comment]
Sounds good to me.

File: src/main/java/seedu/address/model/CommandHistoryModel.java [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+     * Reset the state of {@code tempHistoryList} to be equal to {@code commandHistoryList}.
+     */
+    private void resetTempList() {
+        tempHistoryList = new ArrayList<>(commandHistoryList);
+        tempHistoryList.add(""); // Extra index to store current command.
+        commandHistoryIndex = commandHistoryList.size(); // Reset index to point at current command.
+    }
+
+    /**
+     * Replaces the current command with {@code currenText} and returns the next newest command in the command history.
```
Comment (E0735389) [inline src/main/java/seedu/address/model/CommandHistoryModel.java:44]: Typo "currenText"

File: src/main/java/seedu/address/model/CommandHistoryModel.java [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+     * If the model is already at the newest command, then that command will be returned instead.
+     */
+    public String getNextCommandHistory(String currrentText) {
```
Comment (E0735389) [inline src/main/java/seedu/address/model/CommandHistoryModel.java:47]: Typo "currrentText"

File: src/main/java/seedu/address/model/CommandHistoryModel.java [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+        // Updates command at current index with currentText.
+        tempHistoryList.set(commandHistoryIndex, currrentText);
+
+        commandHistoryIndex = Math.min(commandHistoryIndex + 1, commandHistoryList.size());
+        return tempHistoryList.get(commandHistoryIndex);
+    }
+
+    /**
+     * Replaces the current command with {@code currenText} and returns the next oldest command in the command history.
```
Comment (E0735389) [inline src/main/java/seedu/address/model/CommandHistoryModel.java:56]: Same typo as above (twice)

File: src/main/resources/view/CommandBox.fxml [hunk 4]
```diff
@@ -3,7 +3,6 @@
 <?import javafx.scene.control.TextField?>
 <?import javafx.scene.layout.StackPane?>
 
-<StackPane styleClass="stack-pane" xmlns="http://javafx.com/javafx/17" xmlns:fx="http://javafx.com/fxml/1">
-  <TextField fx:id="commandTextField" onAction="#handleCommandEntered" promptText="Enter command here..."/>
+<StackPane styleClass="stack-pane" xmlns="http://javafx.com/javafx/21" xmlns:fx="http://javafx.com/fxml/1">
```
Comment (E0735389) [inline src/main/resources/view/CommandBox.fxml:6]: You're using different version of JavaFX? (Doesn't sound like a good idea to have incompatible versions...?)

**Review by E0735389 — COMMENTED** [review]
Just minor typo but not that important. These are easy to fix quickly anyway.

**Comment by ChillinRage** [comment]
Sorry I didn't pull the newer changes.

**Comment by ChillinRage** [comment]
I'm not sure if I should add the `requireNonNull(command)` line at CommandHistoryModel line 34 since practically, it's not possible for commandBox to ever call `addCommand` with a null parameter.

**Comment by E0735389** [comment]
Just apply defensive coding in general. The only disadvantage I can think of is to make code larger (marginally, doesn't matter that much) and slow down execution (also marginally, doesn't matter that much)

**Pull request merged**